### ISSUE TITLE
[BUG]: Collision detection does not work below a certain tolerance

### ISSUE THREAD

**Issue by do-wen** [FEniCS/dolfinx#2774]
### How to reproduce the bug

I stumpled across this issue when using the interpolate function for non-matching meshes. When I have only a 'small' shift of coordinates of the two meshes, the interpolate function returns a zero. This behavior is clear for boundary values, but it also happens for values inside both domains. I could track this issue back to the geometry.compute_colliding_cells function.

In the minimal example I presented the issue:
The 'evaluate' function I define is based on this fenicsx tutorial: https://jsdokken.com/dolfinx-tutorial/chapter1/membrane_code.html#making-curve-plots-throughout-the-domain
I am creating two meshes and shift one of those meshes slightly by a random distribution (I added the seed for reproducability). I am then defining some function on mesh 1 and interpolate it to mesh 2. At one position in the function defined on mesh 2 a zero is interpolated, although it is inside both meshes. You can see this clearly in paraview, I also print the problematic DOF to the console. I also write the two functions to paraview where you can have a closed look. Making the eps parameters smaller results in more of these candidates.

Then I am using the 'evaluate' function to further track down the problem (I guess the interpolate function is using something similar). While the cell_candidates are still correct (the bounding box tree finds both elements which might be colliding with the problematic point), the compute_colliding_cells function returns an empty array of cells. I guess because it cannot decide which cell to choose, so it chooses none.

### Minimal Example (Python)

```Python
from dolfinx import io, fem, mesh, geometry
from mpi4py import MPI
import numpy as np

# Evaluate function values as implemented in
# https://jsdokken.com/dolfinx-tutorial/chapter1/membrane_code.html#making-curve-plots-throughout-the-domain
def evaluate_function(points, u):
    domain = u.function_space.mesh
    bb_tree = geometry.BoundingBoxTree(domain, domain.topology.dim)
    cells = []
    points_on_proc = []
    # Find cells whose bounding-box collide with the points
    cell_candidates = geometry.compute_collisions(bb_tree, points)
    print("Cell candidates (still correct):", cell_candidates.links(6166))
    # Choose one of the cells that contains the point
    colliding_cells = geometry.compute_colliding_cells(domain, cell_candidates, points)
    print("Colliding cells (wrong):", colliding_cells.links(6166))
    for i, point in enumerate(points):
        if len(colliding_cells.links(i))>0:
            points_on_proc.append(point)
            cells.append(colliding_cells.links(i)[0])
    points_on_proc = np.array(points_on_proc, dtype=np.float64)
    values = u.eval(points_on_proc, cells)
    return values


file1 = io.XDMFFile(MPI.COMM_WORLD, 'testout/output1.xdmf', 'w')
file2 = io.XDMFFile(MPI.COMM_WORLD, 'testout/output2.xdmf', 'w')

# Create two initially equal meshes
msh1 = mesh.create_rectangle(comm=MPI.COMM_WORLD,
                            points=((0.0, 0.0), (1.0, 1.0)), n=(100, 100),
                            cell_type=mesh.CellType.triangle)
msh2 = mesh.create_rectangle(comm=MPI.COMM_WORLD,
                            points=((0.0, 0.0), (1.0, 1.0)), n=(100, 100),
                            cell_type=mesh.CellType.triangle)
# Shift the geometry of the second mesh slighly in any direction
eps = 5.0e-7
np.random.seed(42)
rand_array = eps * np.random.rand(*msh2.geometry.x.shape)
rand_array[:, -1] = 0.0
msh2.geometry.x[:] += rand_array

V1 = fem.FunctionSpace(msh1, ('CG', 1))
V2 = fem.FunctionSpace(msh2, ('CG', 1))

u1 = fem.Function(V1)
u1.interpolate(lambda x: x[0] ** 2 + x[1]) # Arbitrary function value

u2 = fem.Function(V2)
u2.interpolate(u1)
np.set_printoptions(precision=12)
print("Problematic point of u2 (at index 6166)", msh2.geometry.x[6166])
print("Original point (u1 is defined at)      ", msh1.geometry.x[6166])
print("Magnitude of shift:", np.linalg.norm(msh1.geometry.x[6166]-msh2.geometry.x[6166]))
print("Function value at u2[6166]:", u2.x.array[6166])
print("Function value at u1[6166]:", u1.x.array[6166])

# Test the evaluate function
vals = evaluate_function(msh2.geometry.x, u1)

# Write and close files
file1.write_mesh(msh1)
file1.write_function(u1)
file2.write_mesh(msh2)
file2.write_function(u2)
file1.close()
file2.close()
```

### Version

0.6.0

### DOLFINx git commit

24f86a9ce57df6978070dbee22b3eae8bb77235f

### Installation

I used docker on windows 11

**Comment by jorgensd** [comment]
I can reproduce this with the nightly build with a slightly modified code: 
```python

from dolfinx import io, fem, mesh, geometry, cpp
from mpi4py import MPI
import numpy as np

# Evaluate function values as implemented in
# https://jsdokken.com/dolfinx-tutorial/chapter1/membrane_code.html#making-curve-plots-throughout-the-domain
def evaluate_function(points, u):
    domain = u.function_space.mesh
    bb_tree = geometry.bb_tree(domain, domain.topology.dim)
    cells = []
    points_on_proc = []
    # Find cells whose bounding-box collide with the points
    cell_candidates = geometry.compute_collisions_points(bb_tree, points)
    breakpoint()
    print("Cell candidates (still correct):", cell_candidates.links(0))
    # Choose one of the cells that contains the point
    colliding_cells = geometry.compute_colliding_cells(domain, cell_candidates, points)
    print("Colliding cells (wrong):", colliding_cells.links(0))
    for i, point in enumerate(points):
        if len(colliding_cells.links(i))>0:
            points_on_proc.append(point)
            cells.append(colliding_cells.links(i)[0])
    points_on_proc = np.array(points_on_proc, dtype=np.float64)
    values = u.eval(points_on_proc, cells)
    return values


file1 = io.XDMFFile(MPI.COMM_WORLD, 'output1.xdmf', 'w')
file2 = io.XDMFFile(MPI.COMM_WORLD, 'output2.xdmf', 'w')

# Create two initially equal meshes
N = 50
msh1 = mesh.create_rectangle(comm=MPI.COMM_WORLD,
                            points=((0.0, 0.0), (1.0, 1.0)), n=(N, 50),
                            cell_type=mesh.CellType.triangle)
msh2 = mesh.create_rectangle(comm=MPI.COMM_WORLD,
                            points=((0.0, 0.0), (1.0, 1.0)), n=(N, 50),
                            cell_type=mesh.CellType.triangle)
# Shift the geometry of the second mesh slighly in any direction
eps = 5.0e-7
np.random.seed(42)
rand_array = eps * np.random.rand(*msh2.geometry.x.shape)
rand_array[:, -1] = 0.0
msh2.geometry.x[:] += rand_array
V1 = fem.FunctionSpace(msh1, ('CG', 1))
V2 = fem.FunctionSpace(msh2, ('CG', 1))

u1 = fem.Function(V1)
u1.interpolate(lambda x: x[0] ** 2 + x[1]) # Arbitrary function value
u2 = fem.Function(V2)
data =cpp.fem.create_nonmatching_meshes_interpolation_data(
        u2.function_space.mesh._cpp_object,
        u2.function_space.element,
        u1.function_space.mesh._cpp_object)
u2.interpolate(u1, nmm_interpolation_data=data)
msh2.topology.create_connectivity(msh2.topology.dim-1, msh2.topology.dim)
ext_facets = mesh.exterior_facet_indices(msh2.topology)
ext_dofs = fem.locate_dofs_topological(V2, msh2.topology.dim-1, ext_facets)
interior_dofs = np.ones(len(u2.x.array), dtype=np.bool_)
interior_dofs[ext_dofs] = False
interior_dofs = np.flatnonzero(interior_dofs)
zero_dofs = np.flatnonzero(np.isclose(u2.x.array, 0))

x = V2.tabulate_dof_coordinates()
interior_zero_dofs = np.intersect1d(interior_dofs, zero_dofs)

cells = []
for cell in range(msh2.topology.index_map(msh2.topology.dim).size_local):
    dofs = V2.dofmap.cell_dofs(cell)
    for dof in interior_zero_dofs:
        if dof in dofs:
            cells.append(cell)
cells = np.unique(cells)


print(evaluate_function(x[interior_zero_dofs], u1))

file2.write_mesh(msh2)
file2.write_function(u2)
file1.close()
file2.close()
```
I'll see if I can pin-point what goes wrong, as a pure eval seems to work for me on the parent mesh.

**Comment by jorgensd** [comment]
The collision detection will always only be accurate up to some tolerance.
We use the GJK algorithm (https://en.wikipedia.org/wiki/Gilbert%E2%80%93Johnson%E2%80%93Keerthi_distance_algorithm) which has a limited accuracy (It contains the squared distance as part of its termination requirement). 

It's quite tricky to decide what is a sensible way of choosing a cell and a tolerance (especially in parallel, as a cell might be closer on another process than the current one).

See: #2787 for a workaround that works for your mesh, but is not robust in general

**Comment by jorgensd** [comment]
Resolved with #2788
